### ISSUE TITLE
Optimize GeoHash rendering

### ISSUE THREAD

**Issue by robyngit** [NCEAS/metacatui#2119]
On the `feature-#1720-cesium-data-catalog` branch, geohashes are rendering reasonably quickly now. However there are a couple of remaining issues:
1. When the view is centred on the poles and very zoomed in, too many geohashes render and the map crashes. When the number of Geohashes in the collection exceeds some set amount, we should "down sample" and use the lower precision geohashes instead.
2. Occasionally, after scrolling around the map for awhile, Cesium freezes and then crashes. This could be the same issue as 1 but I suspect something else is going on. Every time we update the geohashes on the map, we remove the old entities and replace them with new ones. If the old ones are not getting removed properly, maybe there is a memory leak?

**Comment by robyngit** [comment]
I've improved the performance a little, but I'm still seeing problem num. 1 described above. It seems that faceting is either the cause of or a contributor to the map crashing when zoomed in at the poles.

**Comment by robyngit** [comment]
🎉 Big breakthrough on this issue to share! After encountering many dead ends, I finally found and fixed the source of the problem.

The issue was not related to faceting or how we were rendering geohashes. Instead, the problem is tied to trying to perform operations with too many geohashes — something we'd only encounter with a 3D map like Cesium and not with Google Maps.

When the camera is heavily tilted, we end up calculating a very low "height" or "altitude." Since we were deciding on a geohash precision level based on a simple mapping of height:precision, we were requesting a high precision of geohashes for these tilted views. The bounding box that we calculated in these views also ends up being large, since the distant background area is still in view. All of this meant that we needed hundreds of thousands of geohashes to cover the extent at the given precision. A similar problem occurs at the poles.

Even just obtaining the hashStrings from the nGeohash library using the bounding box and precision was sometimes extremely slow. Then, when we tried to consolidate these hashStrings by replacing complete groups of 32 geohashes with their parent (1 precision level lower), the browser would often crash.

To address this issue, I've made these changes, among others:
1. Instead of mapping height/altitude in meters to geohash precision, we now calculate the ideal precision based on the bounding box and a maximum threshold of allowed geohashes.
2. We also consolidate hashStrings immediately upon requesting them for a bounding box. We now iterate through the possible precision levels, starting with the least precise, until we find a mix of geohashes that cover the given area. See [this part of the latest commit](https://github.com/NCEAS/metacatui/blob/fa9584b9a7a608229848b1bc00d74d3b2eb573c0/src/js/collections/maps/Geohashes.js#L401-L486) for details.

These changes have been implemented in the `SpatialFilter` model but not in the `CesiumGeohashes` model. With the CesiumGeohashes layer turned off, the spatial filtering is now much faster and stable. I've tested it with a variety of camera views, and I haven't been able to crash the map yet.

Here's what's left to do:
- [ ] Adapt `CesiumGeohashes` to use the new methods, which will be slightly different since we don't want to display consolidated geohashes. I expect the improvements to still be similar.
- [ ] Prevent redundant Solr queries. Currently, we send two requests to Solr whenever the camera placement changes—one from `SpatialFilters` and another from `CesiumGeohashes` (for facet counts). We can combine listeners in one of the connector models to send a single request with each camera change and then pass the results to the respective models.
- [ ] Test all these changes thoroughly, a little bit of code clean up.
- [ ] Update the demo and see if anyone on the team can still manage to crash the map.

*Side note*: I think it would cool if one day we could use high precision geohashes in the foreground, and low precision hashes in the background, getting less precise the further they are from camera. This would add a lot of complexity and isn't required for the next release, but it could be a nice feature in the future.
